### PR TITLE
[Feature](bangc-ops):add roi_align_backward binary op

### PR DESCRIPTION
**Pull request by yuexiaof**
Thanks for your contribution and we appreciate it a lot. 

## 1. Motivation

roi_align_backward二进制适配mlu-ops

## 2. Modification

change bangc-ops/kernels/kernel_wrapper/lib/libextops.a
create bangc-ops/kernels/roi_align_backward/roi_align_backward.cpp
create bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roi_align_backward/roi_align_backward.cpp
create bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roi_align_backward/roi_align_backward.h
create bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roi_align_backward/test_case/case_0.prototxt
create bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roi_align_backward/test_case/case_1.prototxt

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

采用动态阈值测试方案
average mode:200系列标准：diff1;diff2,阈值[10,10];300系列标准diff1;diff2;diff4,阈值[10,10,1]
max mode:200系列标准：diff1;diff2,阈值[10,10];300系列标准diff1;diff2;diff4,阈值[10,10,0.75]

#### 3.1.2 Operator Scheme checklist

|     No.        |                 Details              |            Check Results             |
|----------------|--------------------------------------|--------------------------------------|
|        1       |Supported hardware                    |             MLU290<br>MLU370<br>MLU590         |
|        2       |Job types                             |          U1       |
|        3       |Layouts                               |          NHWC       |
|        4       |Whether multi-dimensions are supported|                                      |
|        5       |Whether element zero is supported     |                                      |
|        6       |Data type(half/float)                 |           half / float           |
|        7       |Whether there is size limit           |                                      |

#### 3.1.3 New Feature Test

If you have checked the following items, please tick the relevant box.

- [ ] Data type test
- [ ] Multi-dimensional tensor test
- [ ] Layout test
- [ ] Different size/integer remainder end segment/alignment misalignment test
- [ ] Zero dimensional tensor test/zero element test
- [ ] stability test
- [ ] Multiple platform test
- [ ] Gen_case module test
- [ ] Nan/INF tests 
- [ ] Bug fix tests
- [ ] For memory leak check details, see[GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [ ] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [ ] For I/O calculation efficiency check details, see: [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md).

#### 3.1.4 Parameter Check

When a new operator is submitted, the test points are given and the test results are stated.

|                   Test Point                    | Acceptance Standard | Test Result (Error Message) |
| ----------------------------------------------- | --------------------| --------------------------- |
| Whether it conforms to the operator restriction |     Normal error    |                             |
| Whether illegal parameters are passed           |     Normal error    |                             |

### 3.2 Accuracy Test

For the cases used in the New Feature Test section, the features and the number of cases are recorded here. When multiple operations are tested, multiple tables are needed to include details of these operations.

Operation:

|Test Point           | Description                      | Quantity |  Comment |
|----------           |----------------------------------|----------|  --------|
|Data type test       |half/float/int8                   |          |          |
|Mult-tensor test     |Supports 1-8 dims                 |          |          |
|Layout test          |Supports NCHW/NHWC                |          |          |
|Zero element test    |Whether to support this test      |          |          |
|Stability test       |--gtest_repeat=NUM<br>--thread=NUM|          |          |
|Mult-platform test   |MLU370/MLU590                     |          |          |
|Nan/INF test         |Whether to support this test      |          |          |
|Memory leak check    |Test result                       |          |          |
|Code coverage check  |Test result                       |          |          |

[^      OK ] /home/yuexiaofan/gpu_test_case/roi_align_backward/roi_align_backward_15_manual_float16_float16_1623154173365.pb
[       OK ] roi_align_backward/TestSuite.mluOp/19 (141 ms)
[----------] 20 tests from roi_align_backward/TestSuite (21096 ms total)

[----------] Global test environment tear-down
[ SUMMARY  ] Total 20 cases of 1 op(s).
ALL PASSED.
[==========] 20 test cases from 1 test suite ran. (22186 ms total)
[  PASSED  ] 20 test cases.

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370-X4

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|-----|----|----|----|-----|
|roi_align_backward|  89  |  167.499  |  0.00459533   |  0.00354086  |  -1  |  float  | input1[10,6,6,16]  input2[10,5] |
|roi_align_backward|  13  |  66.42  |  0.000279447   |  6.61058e-06  |  -1  |  float  | input1[2,2,2,1] boxes[2,5] argmax_x[2,2,2,1] argmax_y[5,7,7,1]|

Platform：MLU590

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|----|----|----|----|-----|
|op_name|    |    |    |    |    |    |     |
|op_name|    |    |    |    |    |    |     |

### 3.4 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

File: bangc-ops/test/mlu_op_gtest/pb_gtest/mlu_op_test_proto/mlu_op_test.proto [hunk 1]
```diff
@@ -271,7 +271,7 @@ message Node {
   optional BallQueryParam  ball_query_param                     = 4008;  // param
   optional FillParam       fill_param                           = 13199;  // param
   optional GetIndicePairsParam get_indice_pairs_param           = 202302031;  // param
-  optional MoeDispatchBackwardDataParam moe_dispatch_backward_data_param = 20230216; // param
+  optional RoialignParam roialign_param                         = 202;  // param
```
Comment (defei-coder) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/mlu_op_test_proto/mlu_op_test.proto:274]: 删除了MoeDispatch相关的参数

File: bangc-ops/test/mlu_op_gtest/pb_gtest/mlu_op_test_proto/mlu_op_test.proto [hunk 2]
```diff
@@ -584,14 +594,6 @@ message FillParam  {
   optional string   value_hex = 4 [default = "0X00000000"];
 }
 
-// param to call mluOpMoeDispatchBackwardData()
```
Comment (defei-coder) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/mlu_op_test_proto/mlu_op_test.proto:587]: 删除了与roiAlignBackward无关的参数，重新检查一下该proto文件，不能直接复制粘贴过来。

File: docs/bangc-docs/user_guide/9_operators/index.rst [hunk 3]
```diff
@@ -663,20 +663,6 @@ Unique
 
 其中 ``x`` 表示输入数据，``y`` 表示输出数据。
 
-.. _moe_dispatch_forward:
```
Comment (defei-coder) [inline docs/bangc-docs/user_guide/9_operators/index.rst:666]: 无关部分被删除了

File: bangc-ops/test/mlu_op_gtest/pb_gtest/mlu_op_test_proto/mlu_op_test.proto [hunk 4]
```diff
@@ -272,7 +272,7 @@ message Node {
   optional BallQueryParam  ball_query_param                     = 4008;  // param
   optional FillParam       fill_param                           = 13199;  // param
   optional GetIndicePairsParam get_indice_pairs_param           = 202302031;  // param
-  optional MoeDispatchBackwardDataParam moe_dispatch_backward_data_param = 20230216; // param
+  optional MoeDispatchBackwardDataParam moe_dispatch_backward_data_param = 20230216; //param
```
Comment (defei-coder) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/mlu_op_test_proto/mlu_op_test.proto:275]: 多了个空格

File: bangc-ops/test/mlu_op_gtest/pb_gtest/mlu_op_test_proto/mlu_op_test.proto [hunk 5]
```diff
@@ -288,11 +288,12 @@ message Node {
   optional IndiceConvolutionBackwardDataParam indice_convolution_backward_data_param = 22120599;  // param
   optional IndiceConvolutionBackwardParam indice_convolution_backward_param = 20230131;  //param
   optional IndiceConvolutionForwardParam indice_convolution_forward_param = 23020721;  //param
-  optional CarafeForwardParam  carafe_forward_param     = 408;  // CarafeForwardParam
-  optional CarafeBackwardParam  carafe_backward_param   = 409;  // CarafeBackwardParam
+  optional CarafeForwardParam  carafe_forward_param             = 408;  // CarafeForwardParam
```
Comment (defei-coder) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/mlu_op_test_proto/mlu_op_test.proto:291]: 还原该算子的空格，尽量不要改动其他算子的内容

**Review by defei-coder — DISMISSED** [review]

File: bangc-ops/mlu_op.h [hunk 6]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
                    const void *boxes,
                    const mluOpTensorDescriptor_t points_indices_desc,
                    void *points_indices);
+// Group:RoiAlign
+/*!
+ * @brief Computes the gradients of images \b grads_image based on the gradients \b grads and
+ * bounding boxes \b boxes to perform the backpropagation of ::mluOpRoiAlign operation. To use
+ * maximum pooling mode or average pooling mode in this operation, call ::mluOpRoiAlignBackward_v2.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:10856]: the maximum 

operation > function

File: bangc-ops/mlu_op.h [hunk 7]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ * Input. A scaling factor that specifies how to map the box coordinates in the origin image to
+ * the coordinates in the output.
+ * @param[in] sampling_ratio
+ * Input. The number of sampling points in the grid used to compute the output.
+ * @param[in] aligned
+ * Input. A boolean value which determines whether to shift the boxes by 0.5 pixel.
+ * @param[in] grads_image_desc
+ * Input. The descriptor of the gradients tensor of the original images.
+ * @param[out] grads_image
+ * Output. Pointer to the MLU memory that stores the gradients tensor of the original images.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:10881]: 对照一下wiki，去掉这些参数描述前面的Input和Output

File: bangc-ops/mlu_op.h [hunk 8]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+/*!
+ * @brief Computes the gradients of images \b grads_image based on the gradients \b grads and
+ * bounding boxes \b boxes to perform the backpropagation of ::mluOpRoiAlign operation. To use
+ * maximum pooling mode or average pooling mode in this operation, call ::mluOpRoiAlignBackward_v2.
+ *
+ * @param[in] handle
+ * Input. Handle to an MLUOP context that is used to manage MLU devices and queues in
+ * the roi_align_backward operation. For detailed information, see ::mluOpHandle_t.
+ * @param[in] grads_desc
+ * Input. The descriptor of the gradient tensor in the backpropagation process. For detailed
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:10862]: The descriptor of the gradient tensor in the backpropagation process. > The descriptor of the tensor \b grads in the backpropagation process.

File: bangc-ops/mlu_op.h [hunk 9]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ * information, see ::mluOpTensorDescriptor_t.
+ * @param[in] grads
+ * Input. Pointer to the MLU memory that stores the gradient tensor.
+ * @param[in] boxes_desc
+ * Input. The descriptor of the bounding boxes tensor. For detailed information, see
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:10867]: The descriptor of the tensor \b boxes.

File: bangc-ops/mlu_op.h [hunk 10]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ * Input. Pointer to the MLU memory that stores the bounding boxes tensor.
+ * @param[in] spatial_scale
+ * Input. A scaling factor that specifies how to map the box coordinates in the origin image to
+ * the coordinates in the output.
+ * @param[in] sampling_ratio
+ * Input. The number of sampling points in the grid used to compute the output.
+ * @param[in] aligned
+ * Input. A boolean value which determines whether to shift the boxes by 0.5 pixel.
+ * @param[in] grads_image_desc
+ * Input. The descriptor of the gradients tensor of the original images.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:10879]: The descriptor of the tensor \b  grads_image of the original images.

File: bangc-ops/mlu_op.h [hunk 11]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ * @param[out] grads_image
+ * Output. Pointer to the MLU memory that stores the gradients tensor of the original images.
+ *
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM, ::MLUOP_STATUS_ARCH_MISMATCH.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:10884]: 去掉句尾的句号

File: bangc-ops/mlu_op.h [hunk 12]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ *
+ * @par Data Type
+ * - This function supports the following data types for gradient tensor \b grads, boxes tensor
+ *   \b boxes, and output tensor \b grads_image. Data type of all tensors should be the same.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:10888]: - The data types of all tensors should be the same.
- The supported data types of input and output tensors are as follows:

File: bangc-ops/mlu_op.h [hunk 13]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ *   - gradient tensor: half, float.
+ *   - boxes tensor: half, float.
+ *   - output tensor: half, float.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:10891]: 这三行去掉句尾的句号

File: bangc-ops/mlu_op.h [hunk 14]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ *
+ * @par Data Layout
+ * - The supported data layout of \b grads, \b boxes, \b grads_images are as follows:
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:10894]: data layouts

File: bangc-ops/mlu_op.h [hunk 15]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ *   - grads tensor: \p MLUOP_LAYOUT_NHWC.
+ *   - boxes tensor: \p MLUOP_LAYOUT_ARRAY, only supports 2D tensor.
+ *   - grads_image tensor: \p MLUOP_LAYOUT_NHWC.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:10897]: 去掉句尾的句号

File: bangc-ops/mlu_op.h [hunk 16]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ * - This function supports the following data types for gradient tensor \b grads, boxes tensor
+ *   \b boxes, and output tensor \b grads_image. Data type of all tensors should be the same.
+ *   - gradient tensor: half, float.
+ *   - boxes tensor: half, float.
+ *   - output tensor: half, float.
+ *
+ * @par Data Layout
+ * - The supported data layout of \b grads, \b boxes, \b grads_images are as follows:
+ *   - grads tensor: \p MLUOP_LAYOUT_NHWC.
+ *   - boxes tensor: \p MLUOP_LAYOUT_ARRAY, only supports 2D tensor.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:10896]: \p MLUOP_LAYOUT_ARRAY, which only supports 2D tensor.

File: bangc-ops/mlu_op.h [hunk 17]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ *   - grads_image tensor: \p MLUOP_LAYOUT_NHWC.
+ *
+ * @par Scale Limitation
+ * - The gradient tensor and output tensor must have four dimensions.
+ * - Size of the fourth dimension of gradient tensor and output tensor must be the same.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:10901]: The size

File: bangc-ops/mlu_op.h [hunk 18]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ * - The bounding boxes tensor \b boxes must have two dimensions.
+ * - Size of the first dimension of gradient tensor and bounding boxes tensor must be the same.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:10903]: The size of

File: bangc-ops/mlu_op.h [hunk 19]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ * - The shape of \b boxes should be [boxes_num, 5].
+ * - \b boxes[i] consists of [image_id, x1, y1, x2, y2]. \p image_id specifies which image this box
+ *   is in, and should be in the range of [0, batch_num - 1]. \p x1 and \p y1 specify the start
+ *   coordinate of this box in origin image. \p x2 and \p y2 specify the end coordinate of this box
+ *   in origin image. \p x1 and \p y1 should be greater than or equal to 0. \p x2 should be greater
+ *   than \p x1. \p y2 should be greater than \p y1.
+ * - \b spatial_scale should be in the range of (0, 1].
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:10910]: 刷一下格式吧，换行有点乱看着

File: bangc-ops/mlu_op.h [hunk 20]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+                      const mluOpTensorDescriptor_t grads_desc,
+                      const void *grads,
+                      const mluOpTensorDescriptor_t boxes_desc,
+                      const void *boxes,
+                      const mluOpTensorDescriptor_t grads_image_desc,
+                      void *grads_image);
+// Group:RoiAlign
+/*!
+ * @brief Computes the gradients of images \b grads_image based on the gradients \b grads,
+ * bounding boxes \b boxes, the coordinate of x axis \b argmax_x and the coordinate of y axis
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:10949]: \b argmax_x and the 》 \b argmax_x, and the

File: bangc-ops/mlu_op.h [hunk 21]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ * \b argmax_y to perform this operation. Compared with ::mluOpRoiAlignBackward, in addition to
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:10950]: operation > function

File: bangc-ops/mlu_op.h [hunk 22]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ * @param[in] aligned
+ * Input. A boolean value which determines whether to shift the boxes by 0.5 pixel. If the value
+ * of \b aligned is set to true, the boxes are shifted by 0.5. If the value of \b aligned is set
+ * to false, the boxes are not shifted.
+ * @param[in] pool_mode
+ * Input. The pooling mode which determines to use maximum pooling mode or average
+ * pooling mode. If the value of \b pool_mode is set to 1, the average pooling mode is used. If
+ * the value of \b pool_mode is set to 0, the maximum pooling mode is used.
+ * @param[in] grads_image_desc
+ * Input. The descriptor of the gradients tensor of the original images. For detailed information,
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:10995]: 这部分参数描述中，去掉Output和Input字样

File: bangc-ops/mlu_op.h [hunk 23]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ * see ::mluOpTensorDescriptor_t.
+ * @param[out] grads_image
+ * Output. Pointer to the MLU memory that stores the \b grads_image tensor .
+ *
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM, ::MLUOP_STATUS_ARCH_MISMATCH.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:11001]: 句尾句号去掉

File: bangc-ops/mlu_op.h [hunk 24]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ *
+ * @par Data Type
+ * - This function supports the following data types for gradient tensor \b grads, boxes tensor
+ *   \b boxes, argmax_x tensor \b argmax_x, argmax_y tensor \b argmax_y and output tensor \b
+ *   grads_image. Data type of all tensors should be the same.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:11006]: - The data types of all tensors should be the same.
- The supported data types of input and output tensors are as follows:

File: bangc-ops/mlu_op.h [hunk 25]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ *   - gradient tensor: half, float.
+ *   - boxes tensor: half, float.
+ *   - argmax_x tensor: half, float.
+ *   - argmax_y tensor: half, float.
+ *   - output tensor: half, float.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:11011]: 这几行句尾句号去掉

File: bangc-ops/mlu_op.h [hunk 26]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ *
+ * @par Data Layout
+ * - The supported data layout of gradient tensor \b grads, boxes tensor \b boxes, argmax_x tensor
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:11014]: data layouts

File: bangc-ops/mlu_op.h [hunk 27]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ *   \b argmax_x, argmax_y tensor \b argmax_y and output tensor \b grads_images are as follows:
+ *   - grads tensor: \p MLUOP_LAYOUT_NHWC.
+ *   - boxes tensor: \p MLUOP_LAYOUT_ARRAY, only supports 2D tensor.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:11017]: which only supports 2D tensor.

其他几行后面的句号去掉

File: bangc-ops/mlu_op.h [hunk 28]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ *   - argmax_x tensor: \p MLUOP_LAYOUT_NHWC.
+ *   - argmax_y tensor: \p MLUOP_LAYOUT_NHWC.
+ *   - grads_image tensor: \p MLUOP_LAYOUT_NHWC.
+ *
+ * @par Scale Limitation
+ * - The gradient tensor \b grads, argmax_x tensor \b argmax_x , argmax_y tensor \b argmax_y and
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:11023]: argmax_y tensor \b argmax_y,

File: bangc-ops/mlu_op.h [hunk 29]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ *   output tensor \b grads_images must have four dimensions.
+ * - Size of the fourth dimension of gradient tensor \b grads, argmax_x tensor \b argmax_x,
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:11025]: The size

File: bangc-ops/mlu_op.h [hunk 30]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ *   argmax_y tensor \b argmax_y and output tensor \b grads_images must be the same.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:11026]: \b argmax_y, and output tensor \b grads_images must be the same.

File: bangc-ops/mlu_op.h [hunk 31]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ * - The bounding boxes tensor \b boxes must have two dimensions.
+ * - Size of the first dimension of gradient tensor \b grads, argmax_x tensor \b argmax_x, argmax_y
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:11028]: The size of the

File: bangc-ops/mlu_op.h [hunk 32]
```diff
@@ -10849,6 +10849,237 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ *   tensor \b argmax_y and bounding boxes tensor \b boxes must be the same.
+ * - Size of each dimension of gradient tensor \b grads, argmax_x tensor \b argmax_x and argmax_y
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:11030]: The size

**Review by GraceGuanH — COMMENTED** [review]
Please check the comments.

File: bangc-ops/mlu_op.h [hunk 33]
```diff
@@ -10849,6 +10849,236 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ *
+ * @par Data Type
+ * - The data types of all tensors should be the same.
+ *   The supported data types of input and output tensors are as follows:
+ *   - gradient tensor: half, float
+ *   - boxes tensor: half, float
+ *   - output tensor: half, float
+ *
+ * @par Data Layout
+ * - The supported data layout of \b grads, \b boxes, and \b grads_images are as follows:
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:10894]: data layouts

File: bangc-ops/mlu_op.h [hunk 34]
```diff
@@ -10849,6 +10849,236 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+ * see ::mluOpTensorDescriptor_t.
+ * @param[out] grads_image
+ * Pointer to the MLU memory that stores the \b grads_image tensor .
+ *
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM, ::MLUOP_STATUS_ARCH_MISMATCH
+ *
+ * @par Data Type
+ * - The data types of all tensors should be the same.
+ *   The supported data types of input and output tensors are as follows
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:11005]: 这一行前面加 -，as follows句尾加英文冒号:

**Review by GraceGuanH — COMMENTED** [review]
Please check the comments.

**Review by GraceGuanH — APPROVED** [review]

**Comment by Goldjin** [comment]
job type 删除 block

**Review by Goldjin — APPROVED** [review]

**Review by Goldjin — APPROVED** [review]

File: docs/bangc-docs/user_guide/9_operators/index.rst [hunk 35]
```diff
@@ -713,3 +713,12 @@ mluOpPointsInBoxes
 	in\_flag = \lvert (z - cz) \rvert <= \frac{dz}{2} \ \& \\
 	\lvert (x - cx) * cos(-rz) - (y - cy) * sin(-rz)\rvert < \frac{dx}{2} \ \& \\
 	\lvert (x - cx) * sin(-rz) + (y - cy) * cos(-rz)\rvert < \frac{dy}{2}
+--------------------------------------
+MoE算法中对输入进行重新分配（dispatch）的反向算子，用于计算gates的梯度'grad_gates'。
```
Comment (duzekunKTH) [inline docs/bangc-docs/user_guide/9_operators/index.rst:717]: 为什么要修改别人的文档部分，而且内容是错的。

File: bangc-ops/mlu_op.h [hunk 36]
```diff
@@ -10849,6 +10849,236 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
...
+                      const float spatial_scale,
+                      const int sampling_ratio,
+                      const bool aligned,
+                      const mluOpTensorDescriptor_t grads_desc,
+                      const void *grads,
+                      const mluOpTensorDescriptor_t boxes_desc,
+                      const void *boxes,
+                      const mluOpTensorDescriptor_t grads_image_desc,
+                      void *grads_image);
+// Group:RoiAlign
```
Comment (duzekunKTH) [inline bangc-ops/mlu_op.h:10946]: 需要空行

File: bangc-ops/mlu_op.h [hunk 37]
```diff
@@ -10849,6 +10849,236 @@ mluOpPointsInBoxes(mluOpHandle_t handle,
                    const void *boxes,
                    const mluOpTensorDescriptor_t points_indices_desc,
                    void *points_indices);
+// Group:RoiAlign
```
Comment (duzekunKTH) [inline bangc-ops/mlu_op.h:10852]: 需要留空行

**Review by Unireverse — APPROVED** [review]

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roi_align_backward/roi_align_backward.h [hunk 38]
```diff
@@ -0,0 +1,54 @@
...
+namespace mluoptest {
+
+class RoiAlignBackwardExecutor : public Executor {
+ public:
+  RoiAlignBackwardExecutor() {}
+  ~RoiAlignBackwardExecutor() {}
+
+  void paramCheck();
+  void compute();
+  void cpuCompute();
```
Comment (tudejiang79) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roi_align_backward/roi_align_backward.h:37]: 以上两个函数加：override;

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roi_align_backward/roi_align_backward.h [hunk 39]
```diff
@@ -0,0 +1,54 @@
...
+  int64_t getTheoryOps() override;
+  void bilinear_interpolate_gradient(int height,
```
Comment (tudejiang79) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roi_align_backward/roi_align_backward.h:39]: 这个是类的私有函数，最好放到private属性里面

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roi_align_backward/roi_align_backward.cpp [hunk 40]
```diff
@@ -0,0 +1,379 @@
...
+#include <iostream>
+#include <vector>
+#include "roi_align_backward.h"
+
+namespace mluoptest {
+
+void RoiAlignBackwardExecutor::paramCheck() {
+  if (!parser_->getProtoNode()->has_roi_align_backward_param()) {
+    LOG(ERROR) << "mluOpRoiAlignBackward: lose param. ";
+  }
```
Comment (tudejiang79) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roi_align_backward/roi_align_backward.cpp:33]: 最好用GTEST_CHECK（）函数进行检查

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roi_align_backward/roi_align_backward.cpp [hunk 41]
```diff
@@ -0,0 +1,378 @@
...
+
+  GTEST_CHECK(
+      parser_->getInputNum() == 2 || parser_->getInputNum() == 4,
+      LOG(ERROR) << "mluOpRoiAlignBackward: tensor input number is wrong.";
+      LOG(ERROR) << "mluOpRoiAlignBackward: tensor input number is "
+                 << parser_->getInputNum());
+  GTEST_CHECK(
+      parser_->getOutputNum() == 1,
+      LOG(ERROR) << "mluOpRoiAlignBackward: tensor output number is wrong.");
+}
```
Comment (tudejiang79) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roi_align_backward/roi_align_backward.cpp:42]: 这里的GTEST_CHECK（）函数不是这样用的吧，你参考下别的算子

**Review by tudejiang79 — APPROVED** [review]

**Pull request merged**